File: clj/src/h3m_lwp_clj/input_processor.clj
(ns h3m-lwp-clj.input-processor
  (:import
   [com.badlogic.gdx Gdx InputProcessor Application$ApplicationType]
   [com.badlogic.gdx.graphics OrthographicCamera]
   [com.badlogic.gdx Input$Keys])
  (:require
   [h3m-lwp-clj.consts :as consts]))


(defn is-desktop?
  []
  (= (.getType (Gdx/app))
     (Application$ApplicationType/Desktop)))


(defn pressed?
  [key _]
  (.isKeyPressed (Gdx/input) key))


(defn create
  [^OrthographicCamera camera]
  (proxy [InputProcessor] []
    (keyDown [keycode] true)
    (keyTyped
      [keycode]
      (when (is-desktop?)
        (condp pressed? keycode
          Input$Keys/NUM_0 (set! (.-zoom camera) 1)
          Input$Keys/EQUALS (set! (.-zoom camera) (- (.-zoom camera) 0.1))
          Input$Keys/MINUS (set! (.-zoom camera) (+ (.-zoom camera) 0.1))
          Input$Keys/UP (.translate camera 0 (- consts/tile-size) 0)
          Input$Keys/DOWN (.translate camera 0 consts/tile-size 0)
          Input$Keys/LEFT (.translate camera (- consts/tile-size) 0 0)
          Input$Keys/RIGHT (.translate camera consts/tile-size 0 0)
          Input$Keys/SPACE (set-random-camera-position camera (:size @h3m-map))
          nil))
      true)
    (keyUp [keycode] true)
    (mouseMoved [x y] true)
    (scrolled [amount] true)
    (touchDown [^Integer screen-x ^Integer screen-y ^Integer pointer ^Integer button] true)
    (touchUp [^Integer screen-x ^Integer screen-y ^Integer pointer ^Integer button] true)
    (touchDragged
      [^Integer screen-x ^Integer screen-y ^Integer pointer]
      (when (is-desktop?)
        (.translate camera (.getDeltaX (Gdx/input)) (.getDeltaY (Gdx/input))))
      true)))

File: core/src/com/heroes3/livewallpaper/core/InputProcessor.kt
package com.heroes3.livewallpaper.core

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.OrthographicCamera
import ktx.app.KtxInputAdapter

class InputProcessor(private val camera: OrthographicCamera) : KtxInputAdapter {
    override fun keyDown(keycode: Int): Boolean {
        when (keycode) {
            Input.Keys.NUM_0 -> camera.zoom = 1f
            Input.Keys.EQUALS -> camera.zoom -= 0.1f
            Input.Keys.MINUS -> camera.zoom += 0.1f
            Input.Keys.UP -> camera.translate(0f, -32f, 0f)
            Input.Keys.DOWN -> camera.translate(0f, 32f, 0f)
            Input.Keys.LEFT -> camera.translate(-32f, 0f, 0f)
            Input.Keys.RIGHT -> camera.translate(32f, 0f, 0f)
        }

        return super.keyDown(keycode)
    }

    override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
        camera.translate(Gdx.input.deltaX.toFloat(), Gdx.input.deltaY.toFloat(), 0f)
        return super.touchDragged(screenX, screenY, pointer)
    }
}